### ISSUE TITLE
Formatting changes

### ISSUE THREAD

**Issue by Billy-Sheppard** [andrewbaxter/genemichaels#29]
Not sure if i just didn't notice before but did it always strip out new lines from the middle of code? I often use a single new line to chunk code into related blocks - rustfmt does respect these but they get flattened with genemichaels.

Secondly this comment: 
```rust
// // find workspace cargo toml
// for parent in 0.. {
//     let dir = if parent == 0 {
//         ".".into()
//     } else {
//         "../".repeat(parent)
//     };
//     match std::fs::read_dir(dir)?.filter_map(|f| f.ok()).find(|f| f.file_name() == "Cargo.toml") {
//         Some(path) if
//             cargo_manifest::Manifest::from_path(path.path())
//                 .ok()
//                 .and_then(|m| m.workspace)
//                 .is_some() => {
//             cargo_toml = Some(path);
//             break;
//         },
//         _ => {
//             continue;
//         },
//     }
// }
```
gets formatted to this:
```rust
// // find workspace cargo toml for parent in 0.. { let dir = if parent == 0 { ".".into() } else {
// "../".repeat(parent) }; match std::fs::read_dir(dir)?.filter_map(|f| f.ok()).find(|f|
// f.file_name() == "Cargo.toml") { Some(path) if cargo_manifest::Manifest::from_path(path.path())
// .ok() .and_then(|m| m.workspace) .is_some() => { cargo_toml = Some(path); break; }, _ => {
// continue; }, } }
```

Is that intended? Seems too aggressive for my liking, would like to hear your thoughts.

EDIT: I do see verbatim comments exist - I feel like there may be a better way to do this, since this is non standard for most users and comment keyboard shortcuts won't natively cover this.

**Comment by andrewbaxter** [comment]
I think there's two cases here, human oriented comments and people commenting out code temporarily, and there's no reliable way to distinguish them (aside from a new comment type like the verbatim comments), so I took the approach of assuming it's human text everywhere.

Comments are treated as human text and reflowed per line length rules in such a way that it'd produce identical markdown visualization, so ` ``` ` can be used to disable text-style reflowing too (although I was planning to hook up rust reformatting for rust-labeled markdown comment code blocks too at some point).

A third option is using config flags to disable code (which is also nice because it keeps formatting + code coloring):
```
#[cfg(all(never))]
```

Basically, reflowing is important because if it doesn't reflow the text it'd need to keep existing formatting in a stable way (i.e. not introduce new spaces or delete old spaces) but that's hard for a lot of reasons.  I'd be happy with making it conditional if that were achieved, but I think this may be one where existing code needs to be updated to work well with this.

**Comment by andrewbaxter** [comment]
FWIW I agree that editor support is an issue.

**Comment by Billy-Sheppard** [comment]
Is it reasonable to attempt to parse the comment as rust code and if its valid source code then don't flatten? This is more of a discussion rather than a `need fix` issue, so I'm interested in working through some thoughts.

Personally I think adding a new comment type is out of scope for a formatter.

**Comment by andrewbaxter** [comment]
I think comments are generally intended for human consumption, and using comments to temporarily disable code is a convenient hack but not an idiomatic usage of comment.  Lots of formatters will tweak comment layouts such that they're not generally great at preserving spacing, lines, etc.

Per the above, I'd like to prioritize formatting of the human comments, so if a normal comment somehow happened to be rust and it formatted it as rust that would be counterproductive.  Also, I think commented out code that's parseable as rust is probably in the minority - a lot of times you'll have only one bracket, or it'll be a commented out partial expression like `.await?` which isn't parseable on its own, etc.

> Thoughts on this?

Oh yeah, that was there from the start.  The output is generated solely from the ast which only contains semantic information, so no formatting is kept.

There are rules to introduce spaces though - I think `fn/struct/enum` definitions get a space before/after, also a comment will get a space before it.

**Comment by Billy-Sheppard** [comment]
Your views on comments seem reasonable, happy to leave the discussion here for now, I wonder if a better solution will come up at some point later anyway, if not I think the current state is fine then.
___
As for keeping stray newlines within code blocks - would it be possible to keep them? I would think it should flatten gaps of more than 1 newline down to 1, however I think removing them all completely is too aggressive. I understand due do how the file is read and parsed this may be impossible though, I wonder if there is a solution. 

I don't know how it works under the hood but I think the ability to retain them is invaluable. From my perspective, `paragraphing` rust code makes it infinitely more readable.

**Comment by andrewbaxter** [comment]
Okay yeah, I'm thinking about this more.  It's definitely not easy, but it's possible we could detect extra blank lines when we're doing comment extraction and re-add them maybe piggybacking on the code to add comments (i.e. add comments and blank lines instead of just comments), although the places new lines could be added is limited to where comments go.

**Comment by Billy-Sheppard** [comment]
I think as long as its stable its very much worth doing. In my mind it's a significant hurdle for people coming from `rustfmt` to adjust to without it.

What it should really come down to is whether you think it's a good feature though, it's a troublesome road to go down if the reason changes are made is "because `rustfmt` does it", obviously.

**Comment by andrewbaxter** [comment]
I think #41 and #53 are more specific and the only remaining discussions here were those, so I'm going to close this.  If there were other topics let's make new specific issues for them and link this.
